### ISSUE TITLE
Investigate duplication of medial items -- does it happen?

### ISSUE THREAD

**Issue by pczarn** [pczarn/gearley#12]
Creation of medial items has only two causes.
- completed item advances a predicted item
- scanned token advances a predicted item

As far as can I see, item completions themselves are unique, so they give unique medial items. Moreover, if scanning was guarded against duplicate completions, we would achieve uniqueness of medial items without the need for deduplication. We still have to sort lists of medial items so that binary search can be used on them later. However, the sorting algorithm can be modified to work faster on lists of unique items.

**Comment by pczarn** [comment]
We do not deduplicate in the current code.

**Comment by Dan-wanna-M** [comment]
I am curious whether item completions themselves are unique. I think this grammar with this string `aaaa` gives a counterexample?
```ebnf
S ::= A A
A ::= 'a'|'aa'
```

**Comment by pczarn** [comment]
@Dan-wanna-M Hi,

A partial parse is a 3-tuple (triple) of `(beginning position .. end position, partial parse symbol)`.

Some partial parses will not appear in the full parse result, and that's okay. In your example, after parsing `a` we get two  partial parses `(0 .. 1, 'a')` `(0 .. 1, A)` which means the input at 0 to 1 was successfully parsed as both `'a'` and `A`. Next, after parsing two more letters we get more partial parses `(1 .. 2, 'a')`, `(1 .. 2, A)`, `(0 .. 2, S)`, `(1 .. 3, A)`, `(0 .. 3, S)`. However, this `(0 .. 3, S)` is not a full parse and will be discarded because the full parse must be cover the entire input like so: `(0 .. 4, S)`.

Is the `(0 .. 3, S)` a problem? No, we handle ambiguity correctly. We just get more than one valid parse.

As you may see, similar completions may appear multiple times in the priority queue, but that's okay, because we run completion only once per `(a .. b, N)` : https://github.com/pczarn/gearley/blob/dcb721ac39c2291269c0d9819ab6ba811a21e013/src/recognizer.rs#L479-L483

We compare `(a .. _, N)` to find duplicates here: https://github.com/pczarn/gearley/blob/dcb721ac39c2291269c0d9819ab6ba811a21e013/src/recognizer.rs#L497-L499

**Comment by pczarn** [comment]
Okay, so in the top comment I said `item completions themselves are unique`, but that is poorly communicated, I meant to say they can be similar but unique by `(a .. b, N)`.

**Comment by Dan-wanna-M** [comment]
> Okay, so in the top comment I said `item completions themselves are unique`, but that is poorly communicated, I meant to say complete items can be similar but completions `(a .. b, N)` run only once per each group of similar items.

So if I understand this correctly, `gearley` still have some kinds of 'item completion deduplication?' 

```txt
-----0-----
S-> .AA (0)
A-> .a (0)
A-> .aa (0)
-----1-----
A->a.(0)
A->a.a (0)
S->A.A(0)
A->.a(1)
A->.aa (1)
-----2-----
A->aa. (0)
A->a.(1)
A->a.a (1)
S-> A.A (0)
S->AA. (0)
A-> .a (2)
A-> .aa (2)
-----3-----
A->aa. (1)
A-> a. (2)
A-> a.a (2)
S->AA. (0) # due to A-> aa. (1)
S-> AA. (0) # due to A-> a. (2)

I am specifically thinking about the general problems characterized by the S->AA.(0) here(assume the string is `aaa`, I think I mistype an extra `a`). According to what you said, you mean the two `S->AA. (0)` will only lead to one completion, if any? And in general cases where S`->AA.C(0)` is still an item, you handle the issue by "if scanning was guarded against duplicate completions"?

**Comment by pczarn** [comment]
Yes, we may call our approach "deduplication" of complete items.

Yes, it will always lead to one completion.  Take your dotted rules `S -> AA.(0)` and grab `S` and `0` from both, compare `0 == 0 && S == S`  they are all equal, respectively. Rule ID is the same, but even if it were not, say for `S -> XY. (0)` and `S -> Z. (0)` the completion still happens once despite rule ID being different.

No, scanning does not need to be guarded, it's unrelated. Think of it like this -- deduplicated completions make unique in-progress items for us. Since completions are deduplicated, and in-progress items come from completions, we do not need to deduplicate them.

**Comment by Dan-wanna-M** [comment]
> Yes, we may call our approach "deduplication" of complete items.
> 
> Yes, it will always lead to one completion. Take your dotted rules `S -> AA.(0)` and grab `S` and `0` from both, compare `0 == 0 && S == S` they are all equal, respectively. Grammar rule is the same, but even if it were not, say for `S -> XY. (0)` and `S -> Z. (0)` the completion still happens once despite grammar rule being different.
> 
> No, scanning does not need to be guarded, it's unrelated.

Got it. But if we have `S -> AA.C(0)` where C is something else instead of `S -> AA.(0)`, wouldn't we have duplicate items?

**Comment by pczarn** [comment]
You made me think a lot on how the parser works under the hood.

Yes, correct, we would have duplicate items if our grammar allows rules of length 3 or more.

I guess binarization prevents us from having any duplicate items. Take a `S -> AAC` and binarize it to `S -> BC`, `B -> AA`.

Now in this case we have twice `B -> AA. (0)` and then once `S -> B.C (0)`. We get only one of the latter thanks to binarization and completion deduplication.  Deduplication of completions is an extra step which prevents duplicate medial items for binarized grammars only.

**Comment by pczarn** [comment]
We binarize rules from the left, by popping two symbols from the **_front_** of the rule and replacing them with an unique generated symbol.

**Comment by pczarn** [comment]
Binarization forces us to do a shitload of completions their and deduplications, but this may be a very good thing since deduplication is implemented once, not twice.

**Comment by pczarn** [comment]
Note how we talk about advancing predicted items in the top comment:

**completed item advances a predicted item**

completion can advance a predicted item, not an in-progress item.

**Comment by pczarn** [comment]
deleted the last comment which was incorrect

**Comment by Dan-wanna-M** [comment]
Got it, thank you for your detailed reply! I am implementing something similar and found myself having to deduplicate both completion and advanced items, thus curious about how you handle the issue.

**Comment by pczarn** [comment]
@Dan-wanna-M 

Your project is some kind of an Earley algorithm, could you explain your goals?

Thanks for asking the right questions needed to close this issue. My answers are the full answer to this topic, so I am closing this issue. The deduplication of scans is a small sub-topic to look into in the future, it may end up in a new issue that I may open later.

You might be interested in the mini-earley I did that showcases my research and allows easy prototyping of new ideas

https://gist.github.com/pczarn/c51dd53b5e194467a79ff5592856d14c

**Comment by pczarn** [comment]
@Dan-wanna-M 

Over the last 8 years, I've been looking into Earley's from time to time and thinking about this algorithm. Another expert in this specialization is mr Jeffrey Kegler, the author of Marpa.

**Comment by Dan-wanna-M** [comment]
> @Dan-wanna-M
> 
> Your project is some kind of an Earley algorithm, could you explain your goals?
> 
> Thanks for asking the right questions needed to close this issue. My answers are the full answer to this topic, so I am closing this issue. The deduplication of scans is a small sub-topic / angle to look into in the future, it may end up in a new issue that I may open later.
> 
> You might be interested in the mini-earley I did that summarizes my research and allows easy prototyping of new ideas
> 
> https://gist.github.com/pczarn/c51dd53b5e194467a79ff5592856d14c

Basically, I am trying to implement constrained decoding engine for language models based on context free grammar, which means I need an incremental recognizer. Earley algorithm then comes handy since it supports full context free grammar and is reasonably efficient on the "practical" grammar. You can find the code here: https://github.com/Dan-wanna-M/kbnf. 

I am currently looking into methods that could make the engine faster on "practical" grammar. The good thing about constrained decoding engine is that I do not really need to parse it; the bad thing is that in the worst case scenario I need to give a throughoutput of at least 150KB bytes/ms, since I have to "lookahead" every token(a byte slice) to see whether it is rejected by the current Earley sets. I currently achieve this in an amortized way with caches, but I hope I need not.

**Comment by Dan-wanna-M** [comment]
> @Dan-wanna-M
> 
> Over the last 8 years, I've been looking into Earley's from time to time and thinking about this algorithm. Another expert in this specialization is mr Jeffrey Kegler, the author of Marpa.

Marpa is definitely very interesting, though I have not had enough time to look into its implementation. I tried to read the paper but was overwhelmed by enormous notations. I am also curious about some of his claims in the paper. For example, I remembered he refers Leo's paper to state Leo optimization enables Earley to parse *LR-Regular* grammar in linear time, but Leo's original paper states every *LR(k)* grammar, which is not the same thing?

**Comment by pczarn** [comment]
`@Dan-wanna-M 

This project, gearley, is essentially Marpa with some bits added and some bits changed. The basic structure (scan / advance / complete) (and Grammar / Recognizer / Event / Parse Result) and thus public API is very similar to Marpa's. Some minor things probably can't and won't be implemented, such as variable-length tokens.

As for `LR-Regular`, I do not quite remember or understand the difference between `LR-Regular` and `LR(k)`, I suppose it has something to do with regular / automata. A quick search on the net tells too little.

All I can say is that AFAIR it is always the case that every `LR(k)` table or grammar can be transformed into `LR(1)` with equivalent table or an equivalent grammar.

There is a really good book D. Grune and Jacobs on these algorithms. Guess I'd have to take a look to know more about `LR-Regular`.

**Comment by pczarn** [comment]
@Dan-wanna-M  

This project, gearley, is essentially Marpa with some bits added and some bits changed. The basic structure (scan / advance / complete) (and Grammar / Recognizer / Event / Parse Result) and thus public API is very similar to Marpa's. Some minor things probably can't and won't be implemented, such as variable-length tokens.

As for `LR-Regular`, I do not quite remember or understand the difference between `LR-Regular` and `LR(k)`, I suppose it has something to do with regular / automata. A quick search on the net tells too little.

All I can say is that AFAIR it is always the case that every `LR(k)` table or grammar can be transformed into `LR(1)` with equivalent table or an equivalent grammar.

There is a really good book D. Grune and Jacobs on these algorithms. Guess I'd have to take a look to know more about `LR-Regular`.

**Comment by pczarn** [comment]
Perhaps automata scans `k` in advance, so it is simply `k`, no?

**Comment by Dan-wanna-M** [comment]
> @Dan-wanna-M
> 
> This project, gearley, is essentially Marpa with some bits added and many bits refactored. The basic structure (scan / advance / complete) (and Grammar / Recognizer / Event / Parse Result) and thus public API is very similar to Marpa's. Some minor things probably can't and won't be implemented, such as variable-length tokens.
> 
> As for `LR-Regular`, I do not quite remember or understand the difference between `LR-Regular` and `LR(k)`, I suppose it has something to do with regular / automata. A quick search on the net tells too little.
> 
> All I can say is that AFAIR it is always the case that every `LR(k)` table or grammar can be transformed into `LR(1)` with equivalent table or an equivalent grammar.
> 
> There is a really good book D. Grune and Jacobs on these algorithms. Guess I'd have to take a look to know more about `LR-Regular`.

Got it, it's good news since I am much more familiar with Rust. By the way, since variable-length tokens will not be implemented, gearley probably will not offer a scanless interface as well?

**Comment by Dan-wanna-M** [comment]
> Perhaps automata scans `k` in advance, so it is simply `k`, no?

https://www.sciencedirect.com/science/article/pii/S0022000073800509/pdf?md5=8ac87046af6d6a710bebfcaebaca8591&pid=1-s2.0-S0022000073800509-main.pdf I took a look at this paper which, I think, invented the concept of `LR-regular.` I think maybe a better word than `regular` is `input-dependent`. Unlike `LR(k)` where `k` is a fixed number for a given grammar, `LR-regular` is `LR(n)` where `n` depends on the input string length rather than on the grammar.

**Comment by pczarn** [comment]
@Dan-wanna-M 

The scanless interface (SLIF) is a big feature and requires plenty of work in the generator https://github.com/pczarn/panini.

The variable length tokens are a minor unrelated feature.

**Comment by pczarn** [comment]
You may be onto something here, because scanless produces variable width tokens, but I believe the implementation for them is much different.

**Comment by pczarn** [comment]
I did not look that much into SLIF and variable-width tokens, all I remember is: SLIF can work on top of two recognizers or some magic with a single recognizer.

**Comment by pczarn** [comment]
@Dan-wanna-M 

I recommend reading gearley, because of Rust, but also because my code is much shorter and simpler MarpaC.

**Comment by pczarn** [comment]
The cfg manipulation crate is pretty big, but that's for nicer grammar form, nicer API and because of feature completeness.

**Comment by pczarn** [comment]
@Dan-wanna-M 

Some numbers

MarpaC has 13,000 lines of code

cfg has 3,848 lines

gearley without forest has 1,400 lines

cfg + gearley - forest have 5,248 lines

cfg + gearley + simple forest should have a total of 5,748 lines
